File: js/lib/experimental/transport/domain/AssetIdentifier.ts
import { ImoNumber } from "../../../ImoNumber";
import { isNullOrWhiteSpace } from "../../../util/util";

export class AssetIdentifier {
    private readonly _tag: number;
    private readonly _imoNumber?: ImoNumber;
    private readonly _otherId?: string;

    private constructor(value: ImoNumber | string) {
        if (typeof value === "string") {
            this._tag = 2;
            this._otherId = value;
        } else {
            this._tag = 1;
            this._imoNumber = value;
        }
    }

    public get imoNumber() {
        return this._tag === 1 ? this._imoNumber : undefined;
    }

    public get otherId() {
        return this._tag === 2 ? this._otherId : undefined;
    }

    public match<T>(
        onImoNumber: (imoNumber: ImoNumber) => T,
        onOtherId: (otherId: string) => T
    ): T {
        switch (this._tag) {
            case 1:
                return onImoNumber(this._imoNumber!);
            case 2:
                return onOtherId(this._otherId!);
            default:
                throw new Error("Tried to match on invalid ShipId");
        }
    }

    public switch(
        onImoNumber: (imoNumber: ImoNumber) => void,
        onOtherId: (otherId: string) => void
    ): void {
        switch (this._tag) {
            case 1:
                onImoNumber(this._imoNumber!);
                break;
            case 2:
                onOtherId(this._otherId!);
                break;
            default:
                throw new Error("Tried to switch on invalid ShipId");
        }
    }

    public toString(): string {
        switch (this._tag) {
            case 1:
                return this._imoNumber!.toString();
            case 2:
                return this._otherId!;
            default:
                throw new Error("Invalid ShipId");
        }
    }

    public static parse(value?: string): AssetIdentifier {
        if (isNullOrWhiteSpace(value))
            throw new Error(`${AssetIdentifier.name}.parse: value is null`);

        let imo: ImoNumber | undefined = undefined;
        if (value.startsWith("IMO") && (imo = ImoNumber.tryParse(value))) {
            return new AssetIdentifier(imo);
        } else {
            return new AssetIdentifier(value);
        }
    }
}

File: csharp/src/Vista.SDK/Experimental/Transport/AssetIdentifier.cs
namespace Vista.SDK.Experimental.Transport;

public readonly record struct AssetIdentifier
{
    private readonly int _tag;
    private readonly ImoNumber _imoNumber;
    private readonly string? _otherId;

    public readonly bool IsImoNumber => _tag == 1;
    public readonly bool IsOtherId => _tag == 2;

    public readonly ImoNumber? ImoNumber => _tag == 1 ? _imoNumber : null;

    public readonly string? OtherId => _tag == 2 ? _otherId : null;

    private AssetIdentifier(ImoNumber value)
    {
        _tag = 1;
        _imoNumber = value;
        _otherId = null;
    }

    private AssetIdentifier(string value)
    {
        _tag = 2;
        _imoNumber = default;
        _otherId = value;
    }

    public readonly T Match<T>(Func<ImoNumber, T> onImoNumber, Func<string, T> onOtherId)
    {
        return _tag switch
        {
            1 => onImoNumber(_imoNumber),
            2 => onOtherId(_otherId!),
            _ => throw new InvalidOperationException("Tried to match on invalid ShipId"),
        };
    }

    public readonly void Switch(Action<ImoNumber> onImoNumber, Action<string> onOtherId)
    {
        if (_tag == 1)
            onImoNumber(_imoNumber);
        else if (_tag == 2)
            onOtherId(_otherId!);
        else
            throw new InvalidOperationException("Tried to switch on invalid ShipId");
    }

    public override readonly string ToString() =>
        _tag switch
        {
            // In ISO-19848, IMO number as ShipID should be prefixed with "IMO"
            1 => _imoNumber.ToString(),
            2 => _otherId!,
            _ => throw new Exception("Invalid state exception"),
        };

    public static AssetIdentifier Parse(ReadOnlySpan<char> value)
    {
        if (value.IsEmpty)
            throw new ArgumentNullException(nameof(value));

        // In ISO-19848, IMO number as ShipID should be prefixed with "IMO"
        if (
            value.StartsWith("IMO".AsSpan(), StringComparison.OrdinalIgnoreCase)
            && SDK.ImoNumber.TryParse(value, out var imo)
        )
            return new AssetIdentifier(imo);
        else
            return new AssetIdentifier(value.ToString());
    }

    public static AssetIdentifier Parse(string value)
    {
        if (value is null)
            throw new ArgumentNullException(nameof(value));

        var span = value.AsSpan();
        // In ISO-19848, IMO number as ShipID should be prefixed with "IMO"
        if (
            span.StartsWith("IMO".AsSpan(), StringComparison.OrdinalIgnoreCase)
            && SDK.ImoNumber.TryParse(span, out var imo)
        )
            return new AssetIdentifier(imo);
        else
            return new AssetIdentifier(value);
    }

    public static implicit operator AssetIdentifier(ImoNumber id) => new AssetIdentifier(id);
}
